### ISSUE TITLE
JavaVersion does not cover all available versions of Java

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#17565]
We're missing Java 11 in 2.1.x. We also need to decide what to do about non-LTS versions (12 and 13) in 2.2.

**Comment by philwebb** [comment]
Java 9 added `Runtime.version()` which returns a `Runtime.Version`(https://docs.oracle.com/javase/10/docs/api/java/lang/Runtime.Version.html) class. I'm tempted to deprecate our enum now and instead add a `featureRelease` attribute to `@ConditionalOnJava` that takes an int.

e.g.:

```java
@ConditionalOnJava(range=OLDER_THAN, featureRelease=13)
```

**Comment by nosan** [comment]
@philwebb 
Maybe it is worth to add a  new annotation?

```java

@Target({ ElementType.TYPE, ElementType.METHOD })
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Conditional(OnJreCondition.class)
public @interface ConditionalOnJre {

	/**
	 * Configures whether the value configured in {@link #value()} shall be considered the
	 * upper exclusive or lower inclusive boundary. Defaults to
	 * {@link Range#EQUAL_OR_NEWER}.
	 * @return the range
	 */
	Range range() default Range.EQUAL_OR_NEWER;

	/**
	 * The {@code java version} to check for. Use {@link #range()} to specify whether the
	 * configured value is an upper-exclusive or lower-inclusive boundary.
	 * @return the java version
	 */
	int value();

	/**
	 * Range options.
	 */
	enum Range {

		/**
		 * Equal to, or newer than the specified {@code java version}.
		 */
		EQUAL_OR_NEWER,

		/**
		 * Older than the specified {@code java version}.
		 */
		OLDER_THAN

	}

}
```

**Comment by wilkinsona** [comment]
I wonder if we should hold off on adding new attributes or annotations for now. In the future I can see us having a `@ConditionalOnRuntimeVersion` that uses the `parse(String)` and `compareTo(Runtime.Version)` methods on `Runtime.version`. While that is a long way off (we'd need a Java 9+ baseline), if what we have today still works it might be nice to get to `@ConditionalOnRuntimeVersion` in as few steps as possible.

**Comment by philwebb** [comment]
We're going to use [this](https://github.com/spring-projects/spring-boot/pull/17549#issuecomment-512835433) suggestion and rework the Javadoc since not all versions are actually supported by us.

**Comment by dreis2211** [comment]
@philwebb Would you be interested in a PR that implements my proposal? Given that you say Boot doesn't support all versions, we're probably talking JDK 12 (because of its CI builds)?

**Comment by wilkinsona** [comment]
Thanks for the offer, @dreis2211. We’re definitely interested in a PR for your proposal. We decided to add 9, 10, 11, 12, and 13 to the enum in both 2.1.x and master. We also want to reword its javadoc so that it does not say “supported”. We should just say “known” or similar instead.

**Comment by dreis2211** [comment]
On my way

**Comment by dreis2211** [comment]
Created the first minimal PR that points to 2.1.x  - I thought this might be easier with the forward merging process. Once you're okay with that and #17590 is merged, I'd like to extend the tests in master via JUnit 5's `@EnabledOnJRE` facilities (another reason why I thought splitting might be wise). Hope you're okay with that approach.

**Comment by snicoll** [comment]
Closing in favour of PR #17590
